### PR TITLE
Core: Fix `Object::has_method()` for script static methods

### LINKED ISSUE #79521: Callables of static methods are not valid if accessed via the class's name
### Godot version

v4.1.stable.official [970459615], v4.1.1.rc1.official [e709ad4d6]

### System information

Ubuntu 23.04

### Issue description

4.1 introduced support for using static methods as callables. However, if the callable is accessed via the class's name, it is considered invalid. This makes using static methods from other classes as callables very difficult.

### Steps to reproduce

```gdscript
class_name Test extends Node

static func static_func() -> void:
	pass

func _ready() -> void:
	print(Test.static_func.is_valid()) # prints false
	print(static_func.is_valid()) # prints true
```

### Minimal reproduction project

[ClassNameCallables.zip](https://github.com/godotengine/godot/files/12062908/ClassNameCallables.zip)

### PR DESCRIPTION
**Pull request by dalexeev**
* Fixes #79521.
* Supersedes #82755.

See https://github.com/godotengine/godot/issues/79521#issuecomment-1636980456.

### CODE REVIEW

**Comment by RandomShaper** [comment]
My take on this very interesting issue:

My guts tell me that `Script::has_method()` belongs to a different concern that `Object::has_method()`. In other words, a script resource has methods appropriate to its nature (dealing with it as a script), so you can't say that a script happening to define some method for the object it will be attached to means that the script resource itself provides that method. In yet other words,  `Object::has_method()` is intended to check the interface of the object itself, while `Script::has_method()` is intended to check the interface of the script inside, only that both happened to get the same name, unfortunately.

Then, there has to be a way for an object to include the interface defined by any attached script in its own, for purposes like calling, property setting/getting, and checking the interface for a method by name. This last point has motivated me to check the current implementation of `Object::has_method()`, to see if it actually includes whatever interface the attached script enrichens it with. For this discussion, these are the relevant lines:
```
	if (script_instance && script_instance->has_method(p_method)) {
		return true;
	}
```

So, script-wise it will only check the methods in the script _if there's an instance_. My first thought was to add code to find a static method in the attached script. But that's not right, because the object the `Callable` is based on, and so it does the validity check on it, is the class, the `Script`.

**Comment by RandomShaper** [comment]
Since I needed to test my hypothesis, I've done the changes and finally PRd them as #82770.

**Comment by dalexeev** [comment]
Yes, I was confused that this method is called the same in `Variant`, `Object` and `Script`.

Here's another solution (virtual multi-level `_has_method()`): https://github.com/dalexeev/godot/commit/6a9944e888e09489aa08dfa96eea5fe561b5e102. But I think that this is overkill, and it [might conflict](https://docs.godotengine.org/en/stable/classes/class_scriptextension.html#class-scriptextension-method-has-method) with GDExtension in the future, if we decide to expose the method.

**Review by RandomShaper — APPROVED** [review]
The current version of this PR includes the best of all the ideas discussed.

Just a note for the next compat breakage: rename `Script::has_(_static)?method()` to `Script::has_script_(_static)?method()` (I think the already existing `Script::has_script_signal()` and others are named like that to disambiguate).

File: core/object/script_language.h [hunk 1]
```diff
@@ -145,6 +145,7 @@ class Script : public Resource {
 	virtual PropertyInfo get_class_category() const;
 #endif // TOOLS_ENABLED
 
+	virtual bool has_static_method(const StringName &p_method) const { return false; }
```
[the remark this answers is not in the document]
Comment (dalexeev) [reply core/object/script_language.h]: Also `GDVIRTUAL_BIND(_has_static_method, "method");`?

**Review by dsnopek — APPROVED** [review]
The GDExtension related changes look good to me! Thanks :-)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**